### PR TITLE
Add latest version of Armadillo

### PR DESCRIPTION
**Pull request by adamjstewart**
I'm having a bit of trouble linking to the Armadillo libraries. When I try compiling something with Armadillo, I'm seeing the following error message:
```
$ g++  -I/soft/spack-0.9.1/opt/spack/linux-centos6-x86_64/gcc-6.1.0/armadillo-7.500.0-inxgdirrm2cqscwwgqagoztainepayo7/include transport_linear_chain.cpp   -o transport_linear_chain
In file included from /soft/spack-0.9.1/opt/spack/linux-centos6-x86_64/gcc-6.1.0/armadillo-7.500.0-inxgdirrm2cqscwwgqagoztainepayo7/include/armadillo:83,
                 from transport_linear_chain.cpp:11:
/soft/spack-0.9.1/opt/spack/linux-centos6-x86_64/gcc-6.1.0/armadillo-7.500.0-inxgdirrm2cqscwwgqagoztainepayo7/include/armadillo_bits/include_superlu.hpp:91:53: error: /blues/gpfs/home/software/spack-0.9.1/opt/spack/1-centos6-x86_64/gcc-6.1.0/superlu-5.2.1-7y4mnsbozswcsktwfvvdreocwobyroy4/include/supermatrix.h: No such file or directory
```
The main thing to note here is that Armadillo thinks `supermatrix.h` is in `opt/spack/1-centos6-x86_64` when it is really in `opt/spack/linux-centos6-x86_64`. Has anyone ever seen something like this before? I contacted a developer, so hopefully they'll have an answer for me. But I assume it's not an RPATH problem on our end.

### CODE REVIEW

**Comment by tgamblin** [comment]
It kind of looks to me like this is in some header or `-D` argument:

`#define linux 1`

Have you tried reordering the includes in `include_superlu.hpp` or `transport_linear_chain.cpp`?

**Comment by adamjstewart** [comment]
Not yet, unfortunately I have zero experience with C++ and only limited experience with C. Maybe I'll play around with it the next time I'm bored. I can see why someone would've created an enum for each operating system. Hopefully the developers will get back to me; it sounds like this wouldn't be an uncommon bug.

**Comment by svenevs** [comment]
Interestingly I was only hit by this with `clang`, gcc and intel had no problems.  It's definitely a macro expansion problem, I reached out to the developers independently with results of both a spack and non-spack (manual) build demonstrating `linux` getting mangled into `1`.  I'll take another look and see if I can find the enum, but debugging other people's macro expansions is ummm.  Well it's basically a nightmare, and this is coming from somebody who does variable C++ template instantiations via macros, object factories by macros, etc.

**Comment by svenevs** [comment]
Wow.  [This is obnoxious](https://stackoverflow.com/questions/19210935/why-does-the-c-preprocessor-interpret-the-word-linux-as-the-constant-1).  Seems that I didn't run into it with GCC 7 because they probably (finally) stopped doing `#define linux 1`.

The reason this is a problem for armadillo is rather contrived.  They gather the directories with `cmake`, but then `configure_file` on a header with the raw paths.

1. Our paths have `linux` in them, which during expansion get mapped to `1`.
2. It will also likely break with spaces in the file paths.

Solution for (1) is `#undef linux`.  @adamjstewart how exactly do I filter files again?  I'd like to do it this way instead of creating a raw patch because I don't know how much this file changes.

I need to change

```
#undef ARMA_INCFILE_WRAP
#define ARMA_INCFILE_WRAP(x) <x>
```

to include an `#undef linux` literally anywhere in the file.  Like it doesn't matter where, it just has to happen.  What's the best way to do this?

**Comment by adamjstewart** [comment]
Hmm, `filter_file` requires you to replace x with y, you can't just add a line easily. It might be better to write a patch (that at least works for whatever versions are in Spack) and contribute it to the developers. They weren't unresponsive, they actually got back to me fairly quickly. They just told me they didn't have time to track down the problem. If they think that's a good enough hack, they'll add it to the next release. If not, I think you've narrowed it down enough for them to come up with something better.

**Comment by svenevs** [comment]
I sent them a different patch, because I'm fairly certain (2) is why they even have the CMake option for disabling that stuff.  It gets pretty complicated, but it's ultimately down to how things get expanded when.  The way that they use the thing we are patching can be changed, and remove the issue entirely.  Since they are using CMake to configure the file, instead of doing what they did I suggested instead just configuring the files directly.

Their approach:
- Configure one file that sets directory paths.  `ARMA_XXX_DIR`
- Elsewhere, use `ARMA_XXX_DIR` and define new files, something like `STR(ARMA_XXX_DIR)/STR(supermatrix.h)`.

Basically, defer the configure and do all of the string processing in cmake first.  It requires large changes, so I sent them what I was able to figure out and told them that hopefully spack people will stop bothering them from me `#undef linux`.  It's generally a bad idea to do that, but it works for me so I'm submitting the patch...which needs review :wink:
